File: crms-backend/src/models/Booking.js
const db = require('../db').db
const { nowIso } = require('../db')
const User = require('./User')
const Vehicle = require('./Vehicle')

function build(row) {
  if (!row) return null
  return {
    _id: String(row.id),
    id: String(row.id),
    customer: row.customer,
    vehicle: row.vehicle,
    pickupDate: row.pickupDate,
    dropoffDate: row.dropoffDate,
    pickupLocation: row.pickupLocation,
    dropoffLocation: row.dropoffLocation,
    totalAmount: row.totalAmount,
    status: row.status,
    specialRequests: row.specialRequests,
    cancellationReason: row.cancellationReason,
    cancellationFee: row.cancellationFee,
    refundAmount: row.refundAmount,
    drivingOption: row.drivingOption,
    driverId: row.driverId,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  }
}

function populate(booking) {
  if (!booking) return null
  const cust = db.prepare('SELECT * FROM users WHERE id = ?').get(booking.customer)
  const veh = db.prepare('SELECT * FROM vehicles WHERE id = ?').get(booking.vehicle)
  booking.customer = User.publicUser(cust)
  booking.vehicle = Vehicle.serialize(veh)
  return booking
}

function getPaymentStatus(bookingId) {
  const p = db.prepare('SELECT status FROM payments WHERE booking = ? ORDER BY createdAt DESC LIMIT 1').get(Number(bookingId))
  if (!p) return 'pending'
  if (p.status === 'completed') return 'paid'
  if (p.status === 'refunded') return 'refunded'
  return 'pending'
}

function toClient(b, populate = false) {
  if (!b) return null
  const customerId = typeof b.customer === 'object' && b.customer ? b.customer.id : b.customer
  const vehicleId = typeof b.vehicle === 'object' && b.vehicle ? b.vehicle.id : b.vehicle
  const pickup = new Date(b.pickupDate)
  const drop = new Date(b.dropoffDate)
  const duration = Math.max(1, Math.ceil((drop - pickup) / (1000 * 60 * 60 * 24)))
  const statusMap = {
    pending: 'upcoming', confirmed: 'upcoming', active: 'upcoming',
    completed: 'completed', cancelled: 'cancelled',
  }
  const obj = {
    id: String(b.id),
    customerId: String(customerId),
    vehicleId: String(vehicleId),
    pickupLocation: b.pickupLocation,
    returnLocation: b.dropoffLocation,
    pickupDate: b.pickupDate ? b.pickupDate.slice(0, 10) : undefined,
    returnDate: b.dropoffDate ? b.dropoffDate.slice(0, 10) : undefined,
    duration,
    drivingOption: b.drivingOption || 'self',
    driverId: b.driverId ?? null,
    vehiclePrice: Math.round((b.totalAmount || 0) / duration),
    driverPrice: 0,
    totalPrice: b.totalAmount,
    status: statusMap[b.status] || b.status,
    paymentStatus: b.paymentStatus || getPaymentStatus(b.id),
    bookingDate: b.createdAt ? b.createdAt.slice(0, 10) : undefined,
    specialRequests: b.specialRequests,
    cancellationReason: b.cancellationReason,
    cancellationFee: b.cancellationFee,
    refundAmount: b.refundAmount,
    createdAt: b.createdAt,
    updatedAt: b.updatedAt,
  }
  if (populate) {
    const veh = typeof b.vehicle === 'object' && b.vehicle ? b.vehicle : Vehicle.serialize(db.prepare('SELECT * FROM vehicles WHERE id = ?').get(b.vehicle))
    const cust = typeof b.customer === 'object' && b.customer ? b.customer : User.publicUser(db.prepare('SELECT * FROM users WHERE id = ?').get(b.customer))
    obj.vehicle = veh ? {
      id: veh.id, name: veh.name, brand: veh.brand, image: veh.image,
      pricePerDay: veh.pricePerDay, category: veh.type, rating: veh.rating,
    } : null
    obj.customer = cust
  }
  return obj
}

function findById(id) {
  return build(db.prepare('SELECT * FROM bookings WHERE id = ?').get(Number(id)))
}

function findMany({ status, customer, limit = 20, offset = 0 }) {
  const where = []
  const vals = []
  if (status) {
    where.push('status = ?')
    vals.push(status)
  }
  if (customer) {
    where.push('customer = ?')
    vals.push(Number(customer))
  }
  const sql = `SELECT * FROM bookings ${where.length ? 'WHERE ' + where.join(' AND ') : ''} ORDER BY createdAt DESC LIMIT ? OFFSET ?`
  vals.push(Number(limit), Number(offset))
  return db.prepare(sql).all(...vals).map(build)
}

function findConflicts(vehicleId, pickupIso, dropoffIso, excludeId) {
  const where = [
    'vehicle = ?',
    "status IN ('pending', 'confirmed', 'active')",
    '((pickupDate <= ? AND pickupDate >= ?) OR (dropoffDate <= ? AND dropoffDate >= ?))',
  ]
  const vals = [Number(vehicleId), dropoffIso, pickupIso, dropoffIso, pickupIso]
  if (excludeId) {
    where.push('id != ?')
    vals.push(Number(excludeId))
  }
  return db.prepare(`SELECT * FROM bookings WHERE ${where.join(' AND ')}`).all(...vals)
}

function create(data) {
  const {
    customer, vehicle, pickupDate, dropoffDate, pickupLocation,
    dropoffLocation, totalAmount, specialRequests, status, drivingOption, driverId,
  } = data
  const ts = nowIso()
  const info = db.prepare(
    `INSERT INTO bookings (customer, vehicle, pickupDate, dropoffDate, pickupLocation, dropoffLocation, totalAmount, status, specialRequests, drivingOption, driverId, cancellationFee, refundAmount, createdAt, updatedAt)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
  ).run(
    Number(customer), Number(vehicle), pickupDate, dropoffDate, pickupLocation,
    dropoffLocation, totalAmount, status || 'pending',
    specialRequests || null, drivingOption || 'self', driverId ? Number(driverId) : null,
    0, 0, ts, ts
  )
  return findById(info.lastInsertRowid)
}

function update(id, fields) {
  const allowed = ['pickupDate', 'dropoffDate', 'pickupLocation', 'dropoffLocation',
    'totalAmount', 'status', 'specialRequests', 'cancellationReason', 'cancellationFee', 'refundAmount',
    'drivingOption', 'driverId']
  const sets = []
  const vals = []
  for (const f of allowed) {
    if (fields[f] !== undefined) {
      sets.push(`${f} = ?`)
      vals.push(fields[f])
    }
  }
  if (!sets.length) return findById(id)
  sets.push('updatedAt = ?')
  vals.push(nowIso())
  vals.push(Number(id))
  db.prepare(`UPDATE bookings SET ${sets.join(', ')} WHERE id = ?`).run(...vals)
  return findById(Number(id))
}

function remove(id) {
  const row = db.prepare('SELECT * FROM bookings WHERE id = ?').get(Number(id))
  db.prepare('DELETE FROM bookings WHERE id = ?').run(Number(id))
  return build(row)
}

module.exports = { build, populate, toClient, findById, findMany, findConflicts, create, update, remove }

File: crms-backend/app/models/booking.py
from app.extensions import db
from datetime import datetime


class Booking(db.Model):
    __tablename__ = 'bookings'

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    vehicle_id = db.Column(db.Integer, db.ForeignKey('vehicles.id'), nullable=False)
    pickup_location = db.Column(db.String(255), nullable=False)
    return_location = db.Column(db.String(255))
    pickup_date = db.Column(db.DateTime, nullable=False)
    return_date = db.Column(db.DateTime)
    driver_option = db.Column(db.Boolean, default=False)
    driver_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=True)
    special_requests = db.Column(db.Text)
    total_amount = db.Column(db.Integer, nullable=False)
    status = db.Column(db.String(50), default='pending')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Customer-facing fields
    dropoff_location = db.Column(db.String(255))
    dropoff_date = db.Column(db.DateTime)
    total_amount_customer = db.Column(db.Float)
    cancellation_reason = db.Column(db.String(255))
    cancellation_fee = db.Column(db.Float, default=0)
    refund_amount = db.Column(db.Float, default=0)
    driving_option = db.Column(db.String(20), default='self')
    trip_id = db.Column(db.Integer, db.ForeignKey('trips.id'), nullable=True)

    user = db.relationship('User', foreign_keys=[user_id], backref='customer_bookings')
    vehicle = db.relationship('Vehicle', backref='bookings')
    driver = db.relationship('User', foreign_keys=[driver_id])
    trip = db.relationship('Trip', backref='booking', uselist=False)

    def to_dict(self):
        pickup = self.pickup_date.isoformat() if self.pickup_date else None
        dropoff = self.dropoff_date.isoformat() if self.dropoff_date else (self.return_date.isoformat() if self.return_date else None)
        total = self.total_amount_customer if self.total_amount_customer is not None else self.total_amount
        duration = 1
        if self.pickup_date and (self.dropoff_date or self.return_date):
            end = self.dropoff_date or self.return_date
            duration = max(1, ((end - self.pickup_date).total_seconds() / 86400))
        return {
            'id': self.id,
            '_id': str(self.id),
            'customerId': self.user_id,
            'vehicleId': self.vehicle_id,
            'pickupLocation': self.pickup_location,
            'returnLocation': self.return_location or self.dropoff_location,
            'pickupDate': pickup,
            'returnDate': dropoff,
            'duration': duration,
            'drivingOption': self.driving_option or ('with_driver' if self.driver_option else 'self'),
            'driverId': self.driver_id,
            'vehiclePrice': round(total / duration) if duration > 0 else total,
            'driverPrice': 0,
            'totalPrice': total,
            'totalAmount': total,
            'status': self.status,
            'paymentStatus': 'pending',
            'bookingDate': self.created_at.isoformat() if self.created_at else None,
            'specialRequests': self.special_requests,
            'cancellationReason': self.cancellation_reason,
            'cancellationFee': self.cancellation_fee,
            'refundAmount': self.refund_amount,
            'createdAt': self.created_at.isoformat() if self.created_at else None,
            'updatedAt': self.updated_at.isoformat() if self.updated_at else None,
        }
